### DISCUSSION TITLE
How can I post XML Data?

### DISCUSSION THREAD

**Discussion by rsoika** [litestar-org/litestar#3330] [General]
Hi, 
I am trying to post data in XML

```python
@dataclass
class Band(BaseModel):
    id: int
    name: str
    genre: str 


@post(path="/demo3", media_type="application/xml")
async def demo3(
    data: Annotated[Band, Body(media_type="application/xml")],
) -> str:
    print("id="+data.id)
    return "ok"
```

But it looks this is not so easy? I get a HTTP Error: `{"status_code":400,"detail":"JSON is malformed: invalid character (byte 0)"}`

It looks like the xml data can not be converted into json. I guess I need some kind of extra code to parse the request body by myself? Can someone help me how to do this?

**Comment by provinzkraut** [comment]
Validation of XML is currently not supported out of the box, so you'll have to handle this yourself.
Currently supported are JSON, MessagePack and multipart/form encoded data.

You can instead receive that raw body via the `body` kwarg and then use that to do your own validation:

```python
@dataclass
class Band(BaseModel):
    id: int
    name: str
    genre: str 


@post(path="/demo3", media_type="application/xml")
async def demo3(
    body: bytes,
) -> str:
    band = parse_xml(body)  # some function that parses your XML
    print("id="+data.id)
    return "ok"
```

**Reply by vkcku** [reply]
We could do something similar as suggested [here](https://github.com/litestar-org/litestar/issues/3272#issuecomment-2026924940) for request encodings as well?

**Reply by vkcku** [reply]
@rsoika you could take a look at [pydantic-xml](https://pydantic-xml.readthedocs.io/) for parsing XML into pydantic models.

**Reply by rsoika** [reply]
Regarding #3272 we have in Jakarta EE a similar concept where a developer can implement custom `MessageBodyReader` an `dMessageBodyWriter` classes and registers them with his Rest API. The Jakarta implementation will then delegate the encoding to this specific implementation. I just mention this as the discussion seems to go in a similar direction...
I will try now parsing the body as you mentioned.

**Reply by iongion** [reply]
The idea of a body writer / reader is great. What it could bring as benefits:

- ability to register custom media types with their own validation, serialization, deserialization
- ability to support multiple media types simultaneously

In .NET a similar pattern is used with body reader/writer and input/output formats and https://learn.microsoft.com/en-us/aspnet/core/web-api/advanced/custom-formatters?view=aspnetcore-9.0

Once registered, an endpoint can harness custom media type in a very elegant way:

```C#
    [HttpPost]
    [Route("test")]
    [Produces("application/json", "application/xml", "applicaton/bson", "application/x-msgpack")]
    [Consumes("application/json", "application/xml", "applicaton/bson", "application/x-msgpack")]
    public async Task<IActionResult> TestWebhook([FromBody] MyCustomPayload payload)
    {
        return Ok(await ProcessInput(payload));
    }
```

the decorators can applied at class level or at method level in controllers, which also solves the problem of having input/ouput formats registered globally, not only at a method level.
